lib/pos_transaction_register.go

```diff
@@ -0,0 +1,655 @@
+
+	return tr.addTransactionNoLock(txn)
+}
+
+func (tr *TransactionRegister) addTransactionNoLock(txn *MempoolTx) error {
+	if txn == nil || txn.Hash == nil {
+		return fmt.Errorf("TransactionRegister.AddTransaction: Transaction or transaction hash is nil")
+	}
+
+	if _, ok := tr.txnMembership[*txn.Hash]; ok {
+		return nil
+	}
+
+	// If the transaction is too large, reject it.
+	if tr.totalTxnsSizeBytes > math.MaxUint64-txn.TxSizeBytes {
+		return fmt.Errorf("TransactionRegister.AddTransaction: Transaction size overflows uint64. Txn size %v, "+
+			"total size %v", txn.TxSizeBytes, tr.totalTxnsSizeBytes)
+	}
+
+	// Determine the min fee of the bucket based on the transaction's fee rate.
+	bucketMinFeeNanosPerKb, bucketMaxFeeNanosPerKB := computeFeeTimeBucketRangeFromFeeNanosPerKB(txn.FeePerKB,
+		tr.minimumNetworkFeeNanosPerKB, tr.feeBucketRateMultiplierBasisPoints)
+	// Lookup the bucket in the map.
+	bucket, bucketExists := tr.feeTimeBucketsByMinFeeMap[bucketMinFeeNanosPerKb]
+	if !bucketExists {
+		// If the bucket doesn't exist, create it and add the transaction to it.
+		bucket = NewFeeTimeBucket(bucketMinFeeNanosPerKb, bucketMaxFeeNanosPerKB)
+	}
+
+	// Add the transaction to the bucket.
+	if err := bucket.AddTransaction(txn); err != nil {
+		return errors.Wrapf(err, "TransactionRegister.AddTransaction: Error adding transaction to bucket: ")
+	}
+
+	if !bucketExists {
+		// If the bucket didn't exist, add it to the set and the map.
+		tr.addBucketNoLock(bucket)
+	}
+
+	tr.totalTxnsSizeBytes += txn.TxSizeBytes
+	tr.txnMembership[*txn.Hash] = txn
+	return nil
+}
+
+// RemoveTransaction removes a transaction from the register. If the transaction does not exist in the register, or its
+// size exceeds the current register size (which should never happen), it is not removed. Returns nil when transaction
+// was successfully removed from the register, or an error otherwise.
+func (tr *TransactionRegister) RemoveTransaction(txn *MempoolTx) error {
+	tr.Lock()
+	defer tr.Unlock()
+
+	return tr.removeTransactionNoLock(txn)
+}
+
+func (tr *TransactionRegister) removeTransactionNoLock(txn *MempoolTx) error {
+	if txn == nil || txn.Hash == nil {
+		return fmt.Errorf("TransactionRegister.RemoveTransaction: Transaction or transaction hash is nil")
+	}
+
+	if _, ok := tr.txnMembership[*txn.Hash]; !ok {
+		return nil
+	}
+
+	if tr.totalTxnsSizeBytes < txn.TxSizeBytes {
+		return fmt.Errorf("TransactionRegister.RemoveTransaction: Transaction with transaction hash %v size %v "+
+			"exceeds total mempool size %v", txn.Hash.String(), txn.TxSizeBytes, tr.totalTxnsSizeBytes)
+	}
+
+	// Determine the min fee of the bucket based on the transaction's fee rate.
+	bucketMinFeeNanosPerKb, _ := computeFeeTimeBucketRangeFromFeeNanosPerKB(txn.FeePerKB,
+		tr.minimumNetworkFeeNanosPerKB, tr.feeBucketRateMultiplierBasisPoints)
+	// Remove the transaction from the bucket.
+	if bucket, exists := tr.feeTimeBucketsByMinFeeMap[bucketMinFeeNanosPerKb]; exists {
```

Comment:
If it doesn't exist thats' actually an error it seems like. Maybe check that case and error?